src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs

```diff
@@ -0,0 +1,131 @@
+
+namespace Riok.Mapperly.Descriptors.Mappings.ExistingTarget;
+
+public class EnsureCapacityInfo
+{
+    public string? SourceAccessor { get; set; }
+    public string TargetAccessor { get; set; } = null!;
+    public TypeSyntax? CollectionType { get; set; }
+    public TypeSyntax? ReadonlyCollectionType { get; set; }
+}
+
+public static class EnsureCapacityBuilder
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+
+    public static bool CanEnsureCapacity(ITypeSymbol sourceType, ITypeSymbol targetType, WellKnownTypes types, out EnsureCapacityInfo? state)
+    {
+        state = null;
+
+        var capacityMethod = targetType.GetMembers(EnsureCapacityName)
+            .OfType<IMethodSymbol>()
+            .FirstOrDefault();
+
+        // if EnsureCapacity is not available then return false.
+        if (capacityMethod is null)
+        {
+            return false;
+        }
+
+        // if target does not have a count then return false.
+        if (!TryGetNonEnumeratedCount(targetType, types, out var targetSizeProperty))
+        {
+            return false;
+        }
+
+        // if source does not have a count/length then get information to get the non enumerated count at runtime.
+        if (!TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+        {
+            sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable);
+
+            var collectionType = ParseTypeName(types.ICollectionT.Construct(iEnumerable!.TypeArguments.ToArray()).ToDisplayString());
+
+            var readonlyCollectionType = ParseTypeName(types.IReadOnlyCollectionT.Construct(iEnumerable!.TypeArguments.ToArray()).ToDisplayString());
+
+            state = new EnsureCapacityInfo()
+            {
+                TargetAccessor = targetSizeProperty,
+                CollectionType = collectionType,
+                ReadonlyCollectionType = readonlyCollectionType
+            };
+            return true;
+        }
+
+        state = new EnsureCapacityInfo()
+        {
+            SourceAccessor = sourceSizeProperty,
+            TargetAccessor = targetSizeProperty
+        };
+
+        return true;
+    }
+
+    public static bool TryBuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target, EnsureCapacityInfo? state, [NotNullWhen(true)] out StatementSyntax? expression)
+    {
+        expression = null;
+        if (state is null)
+        {
+            return false;
+        }
+
+        // if source accessor is avaailable then create simple EnsureCapacity call
+        if (state.SourceAccessor is not null)
+        {
+            expression = EnsureCapacityStatement(target, MemberAccess(ctx.Source, state.SourceAccessor), MemberAccess(target, state.TargetAccessor));
+            return true;
+        }
+
+        // generate check for runtime type, calling EnsureCapacity if it is a collection/readonlyCollection
+        // this is used if source is IEnumerable
```

Comment:
If the target framework supports `Enumerable.TryGetNonEnumeratedCount`, I'd use that instead. This can be found out if the `TryGetNonEnumeratedCount` method symbol is available on the enumerable type.
Probably use a type switch / pattern matching as fallback.

Reply:
I'd thought of doing that but wasn't sure how to check which framework is being used. I could check if Enumerable has the corresponding method.